math/mathmore/src/GSLNLSMinimizer.cxx

```diff
@@ -202,31 +220,36 @@ bool GSLNLSMinimizer::Minimize() {
    std::vector<double> startValues;
 
    // transformation need a grad function.
-   std::unique_ptr<MultiNumGradFunction> gradFunction(new MultiNumGradFunction(*fChi2Func));
-   std::unique_ptr<MinimTransformFunction> trFuncRaw(CreateTransformation(startValues, gradFunction.get()));
+   std::unique_ptr<MultiNumGradFunction> gradFunction;
+   std::unique_ptr<MinimTransformFunction> trFuncRaw;
+   if (!fUseGradFunction) {
+      gradFunction.reset(new MultiNumGradFunction(fitFunc));
+      trFuncRaw.reset(CreateTransformation(startValues, gradFunction.get()));
+   }
+   else {
+      // use pointer stored in BasicMinimizer
+      trFuncRaw.reset(CreateTransformation(startValues));
+   }
    // need to transform in a FitTransformFunction which is set in the residual functions
-   std::unique_ptr<FitTransformFunction> trFunc;
+   std::unique_ptr<FitTransformFunction<Func>> trFunc;
    if (trFuncRaw) {
       //pass ownership of trFuncRaw to FitTransformFunction
-      trFunc.reset(new FitTransformFunction(*fChi2Func, std::move(trFuncRaw)));
-      for (unsigned int ires = 0; ires < fResiduals.size(); ++ires) {
-         fResiduals[ires] = LSResidualFunc(*trFunc, ires);
-      }
-
+      trFunc.reset(new FitTransformFunction<Func>(fitFunc, std::move(trFuncRaw)));
       assert(npar == trFunc->NTot() );
-      // gradFunction is going to be managed by FitTransFormFunction and does not need
-      // to be deleted
+      for (unsigned int ires = 0; ires < size; ++ires) {
+         residualFuncs.push_back(LSResidualFunc<Func>(*trFunc, ires));
+      }
+   }  else {
+      for (unsigned int ires = 0; ires < size; ++ires) {
+        residualFuncs.push_back( LSResidualFunc<Func>(fitFunc, ires) );
+      }
```

Comment:
```suggestion
      for (unsigned int ires = 0; ires < size; ++ires) {
         residualFuncs.emplace_back(*trFunc, ires);
      }
   }  else {
      for (unsigned int ires = 0; ires < size; ++ires) {
        residualFuncs.emplace_back(fitFunc, ires);
      }
```
I think `emplace_back is nicer here, and it doesn't rely on the compiler to optimize the copying